relay.go

```diff
@@ -0,0 +1,111 @@
+package tchannel
+
+import (
+	"errors"
+	"sync"
+)
+
+// RelayHosts allows external wrappers to inject peer selection logic for
+// relaying.
+type RelayHosts interface {
+	// Get returns the host:port of the best peer for the group.
+	Get(group string) string
+	// TODO: add MarkFailed and MarkOK to for feedback loop into peer selection.
+}
+
+type relayItem struct {
+	remapID     uint32
+	destination *Relayer
+}
+
+// A Relayer forwards frames.
+type Relayer struct {
+	sync.RWMutex
+	metrics StatsReporter
+	hosts   RelayHosts
+	items   map[uint32]relayItem
+
+	// Immutable
```

Comment:
I'm honestly not quite sure what you meant by "immutable" here - these all seem like bags of mutable state to me.